client/matching/loadbalancer.go

```diff
@@ -135,6 +140,144 @@ func (lb *defaultLoadBalancer) pickPartition(
+		return &pollToken{fullName: taskQueue.GetName()}
+	}
+
+	parsedName, err := tqname.Parse(taskQueue.GetName())
+	if err != nil || err == nil && !parsedName.IsRoot() {
+		// parse error or partition already picked, use as-is
+		return &pollToken{fullName: taskQueue.GetName()}
+	}
+
+	tqlb := lb.getTaskQueueLoadBalancer(namespaceID, parsedName, taskQueueType)
+
+	// For read path it's safer to return global default partition count instead of root partition, when we fail to
+	// map namespace ID to name.
+	var partitionCount = dynamicconfig.GlobalDefaultNumTaskQueuePartitions
+
+	namespaceName, err := lb.namespaceIDToName(namespaceID)
+	if err == nil {
+		partitionCount = lb.nReadPartitions(string(namespaceName), parsedName.BaseNameString(), taskQueueType)
+	}
+
+	return tqlb.pickReadPartition(partitionCount, lb.forceReadPartition())
+}
+
+func (lb *defaultLoadBalancer) getTaskQueueLoadBalancer(
+	namespaceID namespace.ID, parsedName tqname.Name, tqType enumspb.TaskQueueType,
+) *tqLoadBalancer {
+	key := taskQueueKey{NamespaceID: namespaceID, Name: parsedName, Type: tqType}
+
+	lb.lock.RLock()
+	tqlb, ok := lb.taskQueueLBs[key]
+	lb.lock.RUnlock()
+	if ok {
+		return tqlb
+	}
+
+	lb.lock.Lock()
+	tqlb, ok = lb.taskQueueLBs[key]
+	if !ok {
+		tqlb = newTaskQueueLoadBalancer(key)
+		lb.taskQueueLBs[key] = tqlb
+	}
+	lb.lock.Unlock()
+	return tqlb
+}
+
+func newTaskQueueLoadBalancer(key taskQueueKey) *tqLoadBalancer {
+	return &tqLoadBalancer{
+		taskQueue: key,
+		lock:      sync.Mutex{},
+	}
+}
+
+func (b *tqLoadBalancer) pickReadPartition(partitionCount int, forcedPartition int) *pollToken {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	// ensure we reflect dynamic config change if it ever happens
+	b.ensurePartitionCountLocked(partitionCount)
+
+	partitionID := forcedPartition
+
+	if partitionID < 0 {
+		partitionID = b.pickReadPartitionWithFewestPolls(partitionCount)
+	}
+
+	b.pollerCounts[partitionID]++
+
+	return &pollToken{
+		fullName:    b.taskQueue.Name.WithPartition(partitionID).FullName(),
+		partitionID: partitionID,
+		balancer:    b,
+	}
+}
+
+// caller to ensure that lock is obtained before call this function
+func (b *tqLoadBalancer) pickReadPartitionWithFewestPolls(partitionCount int) int {
+	// pick a random partition to start with
+	startPartitionID := rand.Intn(partitionCount)
+	pickedPartitionID := startPartitionID
+	minPollerCount := b.pollerCounts[pickedPartitionID]
+	for i := 1; i < partitionCount && minPollerCount > 0; i++ {
+		currPartitionID := (startPartitionID + i) % int(partitionCount)
+		if b.pollerCounts[currPartitionID] < minPollerCount {
+			pickedPartitionID = currPartitionID
+			minPollerCount = b.pollerCounts[currPartitionID]
+		}
+	}
+
+	return pickedPartitionID
+}
+
+// caller to ensure that lock is obtained before call this function
+func (b *tqLoadBalancer) ensurePartitionCountLocked(partitionCount int) {
+	if len(b.pollerCounts) == int(partitionCount) {
```

Comment:
```suggestion
	if len(b.pollerCounts) == partitionCount {
```
and similar for the rest of this function

Reply:
addressed in #4981